src/zeta/aca_oam_port_manager.cpp

```diff
@@ -0,0 +1,183 @@
+{
+}
+
+Aca_Oam_Port_Manager::~Aca_Oam_Port_Manager()
+{
+  //clear all oam punt rules
+  for (auto entry : _oam_ports_table) {
+    _delete_oam_ofp(entry.first);
+  }
+  _clear_all_data();
+}
+
+Aca_Oam_Port_Manager &Aca_Oam_Port_Manager::get_instance()
+{
+  // Instance is destroyed when program exits.
+  // It is instantiated on first use.
+  static Aca_Oam_Port_Manager instance;
+  return instance;
+}
+
+// add the OAM punt rule
+int Aca_Oam_Port_Manager::_create_oam_ofp(uint32_t port_number)
+{
+  int overall_rc;
+
+  string opt = "table=0,priority=25,udp,udp_dst=" + to_string(port_number) + ",actions=CONTROLLER";
+
+  overall_rc = ACA_OVS_Control::get_instance().add_flow("br-int", opt.c_str());
+
+  if (overall_rc == EXIT_SUCCESS) {
+    ACA_LOG_INFO("%s", "creat_oam_ofp succeeded!\n");
+  } else {
+    ACA_LOG_ERROR("creat_oam_ofp failed!!! overrall_rc: %d\n", overall_rc);
+  }
+
+  return overall_rc;
+}
+
+// delete the OAM punt rule
+int Aca_Oam_Port_Manager::_delete_oam_ofp(uint32_t port_number)
+{
+  int overall_rc;
+
+  string opt = "udp,udp_dst=" + to_string(port_number);
+
+  overall_rc = ACA_OVS_Control::get_instance().del_flows("br-int", opt.c_str());
+
+  if (overall_rc == EXIT_SUCCESS) {
+    ACA_LOG_INFO("%s", "delete_oam_ofp succeeded!\n");
+  } else {
+    ACA_LOG_ERROR("delete_oam_ofp failed!!! overrall_rc: %d\n", overall_rc);
+  }
+  return overall_rc;
+}
+
+void Aca_Oam_Port_Manager::_clear_all_data()
+{
+  ACA_LOG_DEBUG("%s", "Aca_Oam_Port_Manager::clear_all_data ---> Entering\n");
+
+  // -----critical section starts-----
+  _oam_ports_table_mutex.lock();
+  // All the elements in the unordered_map container are dropped:
+  // their destructors are called, and they are removed from the container,
+  // leaving _oam_ports_table with a size of 0.
+  _oam_ports_table.clear();
+  _oam_ports_table_mutex.unlock();
+  // -----critical section ends-----
+
+  ACA_LOG_DEBUG("%s", "Aca_Oam_Port_Manager::clear_all_data <--- Exiting\n");
+}
+
+// unsafe function, needs to be called inside oam_ports_table_mutex lock
+// this function assumes there is no existing entry for port_number
+void Aca_Oam_Port_Manager::create_entry_unsafe(uint32_t port_number)
+{
+  ACA_LOG_DEBUG("%s", "Aca_Oam_Port_Manager::create_entry_unsafe ---> Entering\n");
+
+  unordered_set<uint32_t> tunnel_ids_table;
+  _oam_ports_table.emplace(port_number, tunnel_ids_table);
+
+  ACA_LOG_DEBUG("%s", "Aca_Oam_Port_Manager::create_entry_unsafe <--- Exiting\n");
+}
+
+// update oam_ports_table and add the OAM punt rule also if this is the first port in the VPC
+void Aca_Oam_Port_Manager::add_vpc(uint32_t port_number, uint32_t tunnel_id)
+{
+  ACA_LOG_DEBUG("%s", "Aca_Oam_Port_Manager::add_vpc ---> Entering\n");
+  // -----critical section starts-----
+  _oam_ports_table_mutex.lock();
+  if (_oam_ports_table.find(port_number) == _oam_ports_table.end()) {
+    create_entry_unsafe(port_number);
+    _create_oam_ofp(port_number);
```

Comment:
_create_oam_ofp can return failure now, let's propagate the failure to this add_vpc function so that it would return an overall_rc.